File: 2-collectors/clojure-collector/src/com/snowplowanalytics/clojure_collector/responses.clj
;;;; Copyright (c) 2012 SnowPlow Analytics Ltd. All rights reserved.
;;;;
;;;; This program is licensed to you under the Apache License Version 2.0,
;;;; and you may not use this file except in compliance with the Apache License Version 2.0.
;;;; You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
;;;;
;;;; Unless required by applicable law or agreed to in writing,
;;;; software distributed under the Apache License Version 2.0 is distributed on an
;;;; "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;;;; See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

;;;; Author:    Alex Dean (mailto:support@snowplowanalytics.com)
;;;; Copyright: Copyright (c) 2012 SnowPlow Analytics Ltd
;;;; License:   Apache License Version 2.0

(ns com.snowplowanalytics.clojure-collector.responses
  "Holds the different HTTP responses sent by clojure-collector"
  (:import (org.apache.commons.codec.binary Base64))
  (:import (org.joda.time DateTime)))

(def pixel-bytes (Base64/decodeBase64 (.getBytes "R0lGODlhAQABAPAAAAAAAAAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="))) ; Can't define ^:const on this as per http://stackoverflow.com/questions/13109958/why-cant-i-use-clojures-const-with-a-java-byte-array
(def pixel (new java.io.ByteArrayInputStream pixel-bytes))
(def ^:const pixel-length (str (alength pixel-bytes)))

(def send-404
  "Respond with a 404"
  {:status  404
   :headers {"Content-Type" "text/plain"}
   :body    (str "404 Not found")})

(def send-200
  "Respond with a 200"
  {:status  200
   :headers {"Content-Type" "text/plain"}
   :body    (str "OK")})

; [cookie-id cookie-duration cookie-contents]
(def send-cookie-and-pixel
  "Responds with a transparent pixel and the cookie"
  {:status  200
   :headers {"Set-Cookie"    (str "sp=" "cookie-id" "; expires=" "[1]" ";" "cookie-contents")
             "P3P"           "policyref=\"/w3c/p3p.xml\", CP=\"NOI DSP COR NID PSA OUR IND COM NAV STA\""
             "Content-Type"  "image/gif"
             "Content-Length" pixel-length}
   :body    pixel})

(defn- set-id-cookie
  "Sets the SnowPlow ID cookie"
  [id duration domain] 
  {:value    (str "sp=" id)
   :expires  (-> (new DateTime) (.plusSeconds duration))
   :domain   domain})

;; TODO: [1] new Date(new Dat`e().getTime()+cookieDuration).toUTCString()

File: 2-collectors/scala-stream-collector/src/main/scala/com.snowplowanalytics.scalacollector/Responses.scala
/* 
 * Copyright (c) 2013-2014 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0, and
 * you may not use this file except in compliance with the Apache License
 * Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Apache License Version 2.0 is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the Apache License Version 2.0 for the specific language
 * governing permissions and limitations there under.
 */

package com.snowplowanalytics.scalacollector

import com.snowplowanalytics.generated.{
  SnowplowEvent,
  TrackerPayload,
  PayloadProtocol,
  PayloadFormat
}
import com.snowplowanalytics.scalacollector.backends._

import java.util.UUID
import org.apache.commons.codec.binary.Base64
//import org.slf4j.LoggerFactory
import spray.http.{DateTime,HttpResponse,HttpEntity,HttpCookie}
import spray.http.HttpHeaders.{`Set-Cookie`,RawHeader}
import spray.http.MediaTypes.`image/gif`

object Responses {
  val pixel = Base64.decodeBase64("R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==")

  def cookie(queryParams: String, requestCookie: Option[HttpCookie],
      userAgent: Option[String], hostname: String, ip: String) = {
    // Use the same UUID if the request cookie contains `sp`.
    val cookieUUID: String =
      if (requestCookie.isDefined) requestCookie.get.content
      else UUID.randomUUID.toString()

    // Construct an event object from the request.

    // TODO: Should the time be in UTC or local?
    // Should the scema make this more clear?
    val timestamp: Long = System.currentTimeMillis / 1000

    val payload = new TrackerPayload(
      PayloadProtocol.Http,
      PayloadFormat.HttpGet,
      queryParams
    )

    val event = new SnowplowEvent(
      timestamp,
      payload,

      // TODO: Should the collector name/version format be more
      // strictly defined in the schema?
      s"${generated.Settings.name}-${generated.Settings.version}",

      // TODO: should we extract the encoding from the queryParams?
      "UTF-8"
    )

    event.hostname = hostname
    event.ipAddress = ip
    if (userAgent.isDefined) event.userAgent = userAgent.get
    // TODO: Not sure if the refererUri can be easily obtained.
    // event.refererUri = 

    // TODO: Use something like:
    // http://spray.io/documentation/1.1-SNAPSHOT/spray-routing/basic-directives/mapHttpResponseHeaders/
    // to map the HttpResponseHeaders into a list.
    // event.headers = 

    // TODO: Is the user ID the cookie we have associated with a user?
    event.userId = cookieUUID

    if (CollectorConfig.backendEnabledEnum == CollectorConfig.Backend.Kinesis) {
      // TODO: What should the key be?
      KinesisBackend.storeEvent(event, "key")
    } else {
      StdoutBackend.printEvent(event)
    }

    // Build the response.
    val responseCookie = HttpCookie(
      "sp", cookieUUID,
      expires=Some(DateTime.now+CollectorConfig.cookieExpiration),
      domain=CollectorConfig.cookieDomain
    )
    val policyRef = CollectorConfig.p3pPolicyRef
    val CP = CollectorConfig.p3pCP
    val headers = List(
      RawHeader("P3P", s"""policyref="${policyRef}", CP="${CP}""""),
      `Set-Cookie`(responseCookie)
    )
    HttpResponse(entity = HttpEntity(`image/gif`, pixel))
      .withHeaders(headers)
  }

  def dump = HttpResponse(entity=KinesisBackend.getRecordsString)

  def notFound = HttpResponse(status = 404, entity = "404 Not found")
  def timeout = HttpResponse(status = 500, entity = s"Request timed out.")
}
